geoportailv3/static/js/mymaps/mymapsdirective.js

```diff
@@ -266,12 +273,89 @@ app.MymapsDirectiveController.prototype.shareMap = function() {
  * @export
  */
 app.MymapsDirectiveController.prototype.closeMap = function() {
-  this.drawnFeatures_.clear();
+  this.drawnFeatures_.clearMymapsFeatures();
+  this.selectedFeatures_.clear();
+};
+
+
+/**
+ * Closes the current anonymous drawing.
+ * @export
+ */
+app.MymapsDirectiveController.prototype.closeAnonymous = function() {
+  this.drawnFeatures_.clearAnonymousFeatures();
   this.selectedFeatures_.clear();
 };
 
 
 /**
+ * Open the dialog to create a new new map from an anoymous drawing.
+ * @export
+ */
+app.MymapsDirectiveController.prototype.openNewMapFromAnonymous = function() {
+  if (!this.appUserManager_.isAuthenticated()) {
+    this.askToConnect();
+  } else {
+    this.creatingFromAnonymous = true;
+  }
+};
+
+
+/**
+ * Add the anonymous drawing into the current map
+ * @export
+ */
+app.MymapsDirectiveController.prototype.addInMymaps = function() {
+  if (!this.appUserManager_.isAuthenticated()) {
+    this.askToConnect();
+  } else {
+    if (this.isMymapsSelected()) {
+      this.drawnFeatures_.copyAnonymousFeaturesToMymaps().then(
+          goog.bind(function(mapinformation) {
+            var mapId = this.appMymaps_.getMapId();
+            var map = {'uuid': mapId};
+            this.onChosen(map);
+          }, this));
+    }
+  }
+};
+
+
+/**
+ * Create a map from an anonymous drawing.
+ * @export
+ */
+app.MymapsDirectiveController.prototype.createMapFromAnonymous = function() {
+  if (!this.appUserManager_.isAuthenticated()) {
+    this.askToConnect();
+  } else {
+    this.appMymaps_.createMap(this.newTitle, this.newDescription,
+        this.newCategoryId).
+        then(goog.bind(function(resp) {
+          if (goog.isNull(resp)) {
+            this.askToConnect();
+          } else {
+            var mapId = resp['uuid'];
+            if (goog.isDef(mapId)) {
+              var map = {'uuid': mapId};
+              this['drawopen'] = true;
+              this.appMymaps_.setMapId(map['uuid']);
+              this.appMymaps_.loadMapInformation().then(
+                  goog.bind(function(mapinformation) {
+                    this.drawnFeatures_.copyAnonymousFeaturesToMymaps().then(
+                        goog.bind(function(mapinformation) {
+                          this.onChosen(map);
+                          var msg = this.gettext_('Carte créée');
+                          this.notify_(msg);
+                          this.creatingFromAnonymous = false;
+                        }, this));
+                  }, this));
+            }}}, this));
+  }
+};
```

Comment:
It looks to me that you could make this code nicer and more readable by using promise chaining. Something like this:

```js
    this.appMymaps_.createMap(...)
        .then(goog.bind(function(resp) {
          if (goog.isNull(resp)) {
            this.askToConnect();
          } else {
            var mapId = resp['uuid'];
            if (goog.isDef(mapId)) {
              var map = {'uuid': mapId};
              this.appMymaps_.setMapId(map['uuid']);
              return this.appMymaps_.loadMapInformation();
            }    
          }
        }), this)
        .then(goog.bind(function(mapInfo) {
          if (goog.isDef(mapInfo)) {
            return this.drawnFeatures_.copyAnonymousFeaturesToMymaps();
          }
          return mapInfo;
        }), this)
        .then(goog.bind(function(mapInfo) {
          ...
        }), this);
```

I'd recommend reading this excellent article: http://www.frangular.com/2012/12/api-promise-angularjs.html (french).